### PR TITLE
INT-195: A/B test for Google Custom Search integration

### PR DESCRIPTION
**Pull request by perjg**
Code to enable showing search results in Mercury via Google Custom Search.

We do this as an A/B test.

Also includes code to make sure the search is limited to the active community.

### CODE REVIEW

File: front/scripts/main/components/SideNavComponent.ts [hunk 1]
```diff
@@ -45,13 +45,28 @@ App.SideNavComponent = Em.Component.extend({
...
 		 */
 		enter(value = ''): void {
-			window.location.assign('%@Special:Search?search=%@&fulltext=Search'.fmt(Mercury.wiki.articlePath, value));
+			var experimentIds = { // Experiment id from Optimizely
+					prod: '3571301500',
+					dev: '3579160288'
+				},
+				variationNumber = Mercury.Utils.VariantTesting.getExperimentVariationNumber(experimentIds);
+
+			if (variationNumber && variationNumber === 2) { // Use Google Search
```
Comment (dianafa) [inline front/scripts/main/components/SideNavComponent.ts]: isn't the `if (variationNumber === 2) ` sufficient here?
Also, let's try do avoid EOL comments.

File: front/scripts/main/routes/ApplicationRoute.ts [hunk 2]
```diff
@@ -149,7 +149,14 @@ App.ApplicationRoute = Em.Route.extend(Em.TargetActionSupport, App.TrackClickMix
 		},
 
 		search: function (searchString : string) {
-			this.transitionTo('searchResults', {queryParams: {q: searchString}});
+			// Only search in current community
+			var basePath : string  = Em.get(Mercury, 'wiki.basePath');
+			basePath = basePath.replace(/.*?:\/\//g, ""); // remove protocol from basePath
```
Comment (dianafa) [inline front/scripts/main/routes/ApplicationRoute.ts]: I would be really nice to create a separate method for removing protocols from paths and write unit test for this! :)

Comment (RafLeszczynski) [inline front/scripts/main/routes/ApplicationRoute.ts]: @perjg do you need just a `location.hostname` or `location.hostname + location. pathname` ?
@rogatty do we have helper method for this in mercury?

Comment (perjg) [inline front/scripts/main/routes/ApplicationRoute.ts]: I need just the base domain name, which is wiki.basePath with the protocol removed.

I couldn't find anywhere with the protocol removed from Em.get(Mercury, 'wiki')

Comment (RafLeszczynski) [inline front/scripts/main/routes/ApplicationRoute.ts]: @perjg If you just need the `hostname` you can get it from `window.location.hostname` without using any RegEx. If this is not available anywhere in Mercury as a helper method or a property, consider adding this so next person who will need the same functionality could reuse it. 

@rogatty: if we don't already have it where would be the best place to add it?

@perjg another thing is the using of RegEx. If you need to so something like this `basePath.replace(/.*?:\/\//g, "");` Please consider moving in to separate method and write unit test for it.

File: front/scripts/main/controllers/SearchResultsController.ts [hunk 3]
```diff
@@ -3,6 +3,7 @@
 'use strict';
 
 App.SearchResultsController = Em.Controller.extend({
-	queryParams: ['q'],
+	queryParams: ['q', 'site'],
```
Comment (rogatty) [inline front/scripts/main/controllers/SearchResultsController.ts]: Why would you use additional query param that keeps the current site url (`http://starwars.wikia.com/search?site=starwars.wikia.com` :see_no_evil:)? Also there is no need to pass it from `ApplicationRoute` as it's accessible from anywhere.

You can put this in `SearchResultsRoute` instead:
```
setupController(controller: any, model: any, transition: EmberStates.Transition): void {
	var site = whateverMethodYouChoose();
	
	this._super(controller, model, transition);
	controller.set('site', site);
}
```

File: front/scripts/main/components/SideNavComponent.ts [hunk 4]
```diff
@@ -45,13 +45,31 @@ App.SideNavComponent = Em.Component.extend({
...
+			// Use Google Search
+			if (variationNumber === 2) {
+				// Hide SideNav
+				this.sendAction('toggleVisibility', false);
+				this.send('searchCancel');
+
+				this.sendAction('search', value);
+				// Use Wikia Search
+			} else {
+				window.location.assign('%@Special:Search?search=%@&fulltext=Search'.fmt(Mercury.wiki.articlePath, value));
```
Comment (Warkot) [inline front/scripts/main/components/SideNavComponent.ts]: You could use ES6 string templates `` `string {$expression}` ``.

File: front/scripts/main/routes/ApplicationRoute.ts (hunk 2 above)
Comment (rogatty) [inline front/scripts/main/routes/ApplicationRoute.ts]: The best place to add generic functions like that would be `/front/scripts/mercury/utils/`.

Altough in this situation I think that `window.location.hostname` is exactly what @perjg needs.

File: front/scripts/main/components/SideNavComponent.ts [hunk 5]
```diff
@@ -45,13 +45,31 @@ App.SideNavComponent = Em.Component.extend({
...
+				variationNumber = Mercury.Utils.VariantTesting.getExperimentVariationNumber(experimentIds);
+
+			// Use Google Search
+			if (variationNumber === 2) {
+				// Hide SideNav
+				this.sendAction('toggleVisibility', false);
+				this.send('searchCancel');
+
+				this.sendAction('search', value);
+				// Use Wikia Search
```
Comment (rogatty) [inline front/scripts/main/components/SideNavComponent.ts]: Please move this comment one line lower, doesn't make much sense here.

File: front/scripts/main/routes/ApplicationRoute.ts (hunk 2 above)
Comment (perjg) [inline front/scripts/main/routes/ApplicationRoute.ts]: @rogatty @RafLeszczynski You're right, what I need is exactly available in window.location.hostname.

I'll just drop the RegEx, but I will keep /front/script/mercury/utils/ in mind as a location for general utility functionality.

**Comment by perjg** [comment]
@rogatty @dianafa @Warkot Is it ok to merge this now?

**Comment by rogatty** [comment]
LGTM

**Pull request merged**